feat: xrpl-evm release doc (#561)
axelarnetwork/axelar-contract-deployments: releases/cosmwasm/2025-02-xrplevm-GMP-v6.4.0.md

Reviewer: File name should be `2025-02-XRPL-EVM-GMP-v6.0.4.md` for consistency. And note that amplifier gateway is deployed from gmp-sdk repo, which is on `v6.0.4`

> # Ripple EVM Sidechain GMP v6.4.0

Reviewer: ```suggestion
# XRPL EVM Sidechain GMP v6.0.4
```
Ripple is the company name

|                | **Owner**                                                                   |
| -------------- | --------------------------------------------------------------------------- |
| **Created By** | @blockchainguyy <ayush@interoplabs.io>                                      |
| **Deployment** | @blockchainguyy <ayush@interoplabs.io>, @talalashraf <talal@interoplabs.io> |

| **Network**          | **Deployment Status** | **Date** |
| -------------------- | --------------------- | -------- |

> | **Devnet Amplifier** | -                     | TBD      |

Reviewer 2: This was deployed (by Peersyst) on February 5

Reviewer: that's only a test connection. We'll be connecting the official testnet to our devnet-amplifier/stagenet as well, similar to how Flow is done

> | **Stagenet**         | -                     | TBD      |

Reviewer 2: We can say the staging environment that we will deploy to testnet will be `stagenet` for this purpose

Reviewer: That's only temporary, we'll be adding the proper stagenet connection for long term support. We are skipping it temporarily to get the staging connection on testnet. We can update this doc before the official testnet deployment

| **Testnet**          | -                     | TBD      |
| **Mainnet**          | -                     | TBD      |

- [Amplifier Releases](https://github.com/axelarnetwork/axelar-amplifier/releases)
- [VotingVerifier v1.1.0](https://github.com/axelarnetwork/axelar-amplifier/releases/tag/voting-verifier-v1.1.0)
- [Gateway v1.1.1](https://github.com/axelarnetwork/axelar-amplifier/releases/tag/gateway-v1.1.1)
- [MultisigProver v1.1.1](https://github.com/axelarnetwork/axelar-amplifier/releases/tag/multisig-prover-v1.1.1)

## Background

> These are the instructions for deploying Amplifier contracts for Ripple EVM connection.

Reviewer 2: ```suggestion
These are the instructions for deploying Amplifier contracts for XRPL EVM Sidechain connection.
```

### Pre-requisites

Predict the [External Gateway](../evm/2025-02-xrplevm-GMP-v6.4.0.md) address, as `VotingVerifier` needs the `sourceGatewayAddress` which is the External Gateway address.

| Network              | `minimumRotationDelay` | `deploymentType` | `deployer`                                   |
| -------------------- | ---------------------- | ---------------- | -------------------------------------------- |
| **Devnet-amplifier** | `0`                    | `create3`        | `0xba76c6980428A0b10CFC5d8ccb61949677A61233` |

> | **Stagenet**         | `300`                  | `create3`        | `0xba76c6980428A0b10CFC5d8ccb61949677A61233` |

Reviewer 2: To repeat my comment from above, our staging environment on testnet can be considered `stagenet` here. So yes, apporpriate to use 0xba76c6980428A0b10CFC5d8ccb61949677A61233

> | **Testnet**          | `3600`                 | `create`         | `0xB8Cd93C83A974649D76B1c19f311f639e62272BC` |

Reviewer: staging connection shouldn't use the official testnet key `0xB8Cd93C83A974649D76B1c19f311f639e62272BC` otherwise it'll consume the nonce, and we can't derive the official gateway address. Use `0xba76c6980428A0b10CFC5d8ccb61949677A61233` with a different salt

| **Devnet-amplifier** | `axelar1zlr7e5qf3sz7yf890rkh9tcnu87234k6k7ytd9` | `axelar1zlr7e5qf3sz7yf890rkh9tcnu87234k6k7ytd9` |
| **Stagenet**         | `axelar10d07y265gmmuvt4z0w9aw880jnsr700j7v9daj` | `axelar1l7vz4m5g92kvga050vk9ycjynywdlk4zhs07dv` |
| **Testnet**          | `axelar10d07y265gmmuvt4z0w9aw880jnsr700j7v9daj` | `axelar17qafmnc4hrfa96cq37wg5l68sxh354pj6eky35` |
| **Mainnet**          | `axelar10d07y265gmmuvt4z0w9aw880jnsr700j7v9daj` | `axelar1pczf792wf3p3xssk4dmwfxrh6hcqnrjp70danj` |

| Network              | `serviceName` | `votingThreshold` | `signingThreshold` | `confirmationHeight` |
| -------------------- | ------------- | ----------------- | ------------------ | -------------------- |
| **Devnet-amplifier** | `validators`  | `["6", "10"]`     | `["6", "10"]`      | `1`                  |
| **Stagenet**         | `amplifier`   | `["51", "100"]`   | `["51", "100"]`    | `1`                  |
| **Testnet**          | `amplifier`   | `["51", "100"]`   | `["51", "100"]`    | `1`                  |
| **Mainnet**          | `amplifier`   | `["2", "3"]`      | `["2", "3"]`       | `1`                  |

```bash

> # Add under `config.axelar.contracts.VotingVerifier` based on Network

Reviewer 2: When we add the staging env on testnet, should we use `xrplevmtest`?

5. Register `xrplevm` gateway at the Router

```bash
node cosmwasm/submit-proposal.js execute \
  -c Router \
  -t "Register Gateway for xrplevm" \
  -d "Register Gateway address for xrplevm at Router contract" \
  --runAs $RUN_AS_ACCOUNT \
  --deposit $DEPOSIT_VALUE \
  --msg "{
    \"register_chain\": {
      \"chain\": "xrplevm",
      \"gateway_address\": \"$GATEWAY\",
      \"msg_id_format\": \"hex_tx_hash_and_event_index\"
      }
    }"
```

```bash
axelard q wasm contract-state smart <router-address> '{"chain_info": "xrplevm"}' --output json | jq .
# You should see something like this:
{
  "data": {
    "name": "xrplevm",
    "gateway": {
      "address": "axelar1hzz0s0ucrhdp6tue2lxk3c03nj6f60qy463we7lgx0wudd72ctmsee8enx"
    },
    "frozen_status": 0,
    "msg_id_format": "hex_tx_hash_and_event_index"
  }
}
```

6. Update ampd with the `xrplevm` chain configuration. Verifiers should use their own `xrplevm` RPC node for the `http_url` in production.

| Network              | `http_url`              |
| -------------------- | ----------------------- |
| **Devnet-amplifier** | https://rpc.xrplevm.org |

> | **Stagenet**         | https://rpc.xrplevm.org |

Reviewer 2: ```suggestion
| **Stagenet**         | https://rpc.testnet.xrplevm.org |
```

> | **Testnet**          | https://rpc.xrplevm.org |

Reviewer 2: ```suggestion
| **Testnet**          | https://rpc.testnet.xrplevm.org |
```

| **Mainnet**          | ``                      |

```bash
[[handlers]]
chain_finalization="RPCFinalizedBlock"
chain_name="xrplevm"
chain_rpc_url=[http url]
cosmwasm_contract="$VOTING_VERIFIER"
type="EvmMsgVerifier"

[[handlers]]
chain_finalization="RPCFinalizedBlock"
chain_name="xrplevm"
chain_rpc_url=[http url]
cosmwasm_contract="$VOTING_VERIFIER"
type="EvmVerifierSetVerifier"
```

7. Update ampd with the `xrplevm` chain configuration.

```bash

> ampd register-chain-support "[service name]" xrplevm

Reviewer 3: Do we need to register public key before this step: "ampd register-public-key ed25519"?

      \"chain_name\": "xrplevm",
      \"new_prover_addr\": \"$MULTISIG_PROVER\"
    }
  }"
```

9. Authorize `xrplevm` Multisig prover on Multisig

```bash
node cosmwasm/submit-proposal.js execute \
  -c Multisig \
  -t "Authorize Multisig Prover for xrplevm" \
  -d "Authorize Multisig Prover address for xrplevm at Multisig contract" \
  --runAs $RUN_AS_ACCOUNT \
  --deposit $DEPOSIT_VALUE \
  --msg "{
    \"authorize_callers\": {
      \"contracts\": {
        \"$MULTISIG_PROVER\": "xrplevm"
      }
    }
  }"
```

```bash
axelard q wasm contract-state smart <multisig-addr> '{"is_caller_authorized": {"contract_address": "<xrplevm-multisig-prover-addr>", "chain_name": "xrplevm"}}' --output json | jq .
# Result should look like:
{
  "data": true
}
```

10. Create reward pool for voting verifier

#### Rewards

| Network              | `epoch_duration` | `participation_threshold` | `rewards_per_epoch` |
| -------------------- | ---------------- | ------------------------- | ------------------- |
| **Devnet-amplifier** | `100`            | `["7", "10"]`             | `100`               |
| **Stagenet**         | `600`            | `["7", "10"]`             | `100`               |

> | **Testnet**          | `14845`          | `["7", "10"]`             | `100`               |

Reviewer 2: @talalashraf do you think we should use a smaller value than `14845` for testnet for testing purposes?

        \"rewards_per_epoch\": [rewards per epoch]
      },
      \"pool_id\": {
        \"chain_name\": "xrplevm",
        \"contract\": \"$MULTISIG\"
      }
    }
  }"
```

12. Add funds to reward pools from a wallet containing the reward funds `$REWARD_AMOUNT`

```bash
axelard tx wasm execute $REWARDS "{ \"add_rewards\": { \"pool_id\": { \"chain_name\": "xrplevm", \"contract\": \"$MULTISIG\" } } }" --amount $REWARD_AMOUNT --from $WALLET
axelard tx wasm execute $REWARDS "{ \"add_rewards\": { \"pool_id\": { \"chain_name\": "xrplevm", \"contract\": \"$VOTING_VERIFIER\" } } }" --amount $REWARD_AMOUNT --from $WALLET
```

13. Create genesis verifier set

Note that this step can only be run once a sufficient number of verifiers have registered.

| Network              | `min_num_verifiers` |
| -------------------- | ------------------- |
| **Devnet-amplifier** | 3                   |
| **Stagenet**         | 3                   |
| **Testnet**          | 21                  |
| **Mainnet**          | 25                  |

```bash
axelard tx wasm execute $MULTISIG_PROVER '"update_verifier_set"' --from $PROVER_ADMIN --gas auto --gas-adjustment 1.2
```

## Checklist

> The [xrplevm GMP checklist](../ecm/2025-02--xrplevm-GMP-v6.4.0.md) will test GMP call.

Reviewer 3: typo on name path: ecm -> evm?
